### PR TITLE
Implement Zalesak Correction

### PR DESCRIPTION
**Pull request by Zhengyu-Huang**
- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

The advection test with periodic boundary condition for 1 period 
exact_and_computed_advected_square_wave_ZalesakFCT

The advection test with non-penetration boundary condition
exact_and_computed_advected_square_wave_ZalesakFCT

### CODE REVIEW

File: examples/Project.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
 ProgressLogging = "33c8b6b6-d38a-422a-b730-caa89a2f386c"
 QuadGK = "1fd47b50-473d-5c70-9696-f719f8f3bcdc"
 SciMLBase = "0bca4576-84f4-4d90-8ffe-ffa030f20462"
+SnoopPrecompile = "66db9d55-30c0-4569-8b51-7e840670fc0c"
```
Comment (valeriabarra) [inline examples/Project.toml:27]: The `examples/Manifest.toml` and `examples/Project.toml` files shouldn't have been checked-in, unless this test adds a specific new necessary dependency

File: examples/Manifest.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # This file is machine-generated - editing it directly is not advised
 
-julia_version = "1.7.3"
+julia_version = "1.7.1"
```
Comment (charleskawczynski) [inline examples/Manifest.toml:3]: @Zhengyu-Huang, it looks like you're using an older version of Julia, can you upgrade? Let's revert these manifest file changes

File: examples/Project.toml (hunk 1 above)
Comment (charleskawczynski) [inline examples/Project.toml:27]: Let's remove this

File: src/Operators/finitedifference.jl [hunk 3]
```diff
@@ -1592,6 +1592,214 @@ end
...
+Φ_j^{n+1} = Φ_j^{td} - \frac{dt}{dx}(C_{j+\frac{1}{2}}A_{j+\frac{1}{2}} - C_{j-\frac{1}{2}}A_{j-\frac{1}{2}})
+```
+This stencil is based on [durran2013numerical](@cite).
+
+Supported boundary conditions are:
+- [`ThirdOrderOneSided(x₀)`](@ref): uses the third-order downwind reconstruction to compute `x` on the left boundary,
+and the third-order upwind reconstruction to compute `x` on the right boundary.
+
+!!! note
+    Similar to the [`Upwind3rdOrderBiasedProductC2F`](@ref) operator, these boundary conditions do not define the value at the actual boundary faces, and so this operator cannot be materialized directly: it needs to be composed with another operator that does not make use of this value, e.g. a [`DivergenceF2C`](@ref) operator, with a [`SetValue`](@ref) boundary.
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:1617]: Doc string needs fixed, also, can we break the line so that it doesn't go so far out?

File: src/Operators/finitedifference.jl [hunk 4]
```diff
@@ -1592,6 +1592,214 @@ end
...
+    Rⱼ⁻ = (Pⱼ⁻ > 0 ? min(1, Qⱼ⁻ / Pⱼ⁻) : 0)
+
+    ϕⱼ₊₁ᵐᵃˣ = max(ϕⱼ, ϕⱼ₊₁, ϕⱼ₊₂, ϕⱼᵗᵈ, ϕⱼ₊₁ᵗᵈ, ϕⱼ₊₂ᵗᵈ)
+    ϕⱼ₊₁ᵐⁱⁿ = min(ϕⱼ, ϕⱼ₊₁, ϕⱼ₊₂, ϕⱼᵗᵈ, ϕⱼ₊₁ᵗᵈ, ϕⱼ₊₂ᵗᵈ)
+    Pⱼ₊₁⁺ = max(0, Aⱼ₊₁₂) - min(0, Aⱼ₊₃₂)
+    Qⱼ₊₁⁺ = (ϕⱼ₊₁ᵐᵃˣ - ϕⱼ₊₁ᵗᵈ) * dxⱼ₊₁
+    Rⱼ₊₁⁺ = (Pⱼ₊₁⁺ > 0 ? min(1, Qⱼ₊₁⁺ / Pⱼ₊₁⁺) : 0)
+    Pⱼ₊₁⁻ = max(0, Aⱼ₊₃₂) - min(0, Aⱼ₊₁₂)
+    Qⱼ₊₁⁻ = (ϕⱼ₊₁ᵗᵈ - ϕⱼ₊₁ᵐⁱⁿ) * dxⱼ₊₁
+    Rⱼ₊₁⁻ = (Pⱼ₊₁⁻ > 0 ? min(1, Qⱼ₊₁⁻ / Pⱼ₊₁⁻) : 0)
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:1682]: Please fix the type instabilities here, see the julia docs on [type promotions](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/).

File: src/Operators/finitedifference.jl [hunk 5]
```diff
@@ -1592,6 +1592,214 @@ end
...
+    Φᵗᵈ_space,
+)
+    face_space = axes(A_space)
+    @assert idx <= right_face_boundary_idx(face_space) - 1
+
+    Aⱼ₊₁₂ = Geometry.contravariant3(
+        getidx(A_space, loc, idx, hidx),
+        Geometry.LocalGeometry(face_space, idx, hidx),
+    )
+    return Geometry.Contravariant3Vector(Aⱼ₊₁₂ * 0.0)
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:1798]: Why is this multiplying by zero?

File: src/Operators/finitedifference.jl [hunk 6]
```diff
@@ -1592,6 +1592,214 @@ end
...
+    Φᵗᵈ_space,
+)
+    face_space = axes(A_space)
+    @assert idx <= left_face_boundary_idx(face_space) + 1
+
+    Aⱼ₊₁₂ = Geometry.contravariant3(
+        getidx(A_space, loc, idx, hidx),
+        Geometry.LocalGeometry(face_space, idx, hidx),
+    )
+    return Geometry.Contravariant3Vector(Aⱼ₊₁₂ * 0)
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:1778]: Why is this multiplying by zero?

File: docs/refs.bib [hunk 7]
```diff
@@ -30,6 +30,14 @@ @article{BorisBook1973
   author = {Jay P Boris and David L Book}
 }
 
+@book{durran2013numerical,
```
Comment (valeriabarra) [inline docs/refs.bib:33]: I would also add the original citation, Zalesak 1979

File: src/Operators/finitedifference.jl (hunk 5 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl:1798]: I want to set to 0, but if I set it to 0, I got errors

File: src/Operators/finitedifference.jl (hunk 4 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl]: I change all 0 to `stable_zero = zero(eltype(Aⱼ₊₁₂))`

File: src/Operators/finitedifference.jl (hunk 3 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl]: fixed

File: src/Operators/finitedifference.jl (hunk 5 above)
Comment (valeriabarra) [reply src/Operators/finitedifference.jl:1801]: Does A here represent the same thing as in the BB FCT? If so, in #888 we left only the value of A in the BCs. Why here we want this to be always zero?

File: src/Operators/finitedifference.jl [hunk 8]
```diff
@@ -1592,6 +1592,217 @@ end
     return Geometry.Contravariant3Vector(vᶠ * a)
 end
 
+
+
+#########################
+"""
+    U = FCTZalesak(;boundaries)
+    U.(v, x)
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:1600]:
```suggestion
    U.(A, Φ, Φᵗᵈ)
```
?

File: src/Operators/finitedifference.jl (hunk 5 above)
Comment (charleskawczynski) [reply src/Operators/finitedifference.jl:1801]: If it should be zero, then use `zero` (first determine the correct type), and don't call `contravariant3` (this performs operations). If it's not meant to be zero, then let's correct that issue

File: examples/column/zalesak_fct_advection.jl [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+mesh = Meshes.IntervalMesh(domain, nelems = n)
+cs = Spaces.CenterFiniteDifferenceSpace(mesh)
+fs = Spaces.FaceFiniteDifferenceSpace(cs)
+zc = Fields.coordinate_field(cs)
+
+# Unitary, constant advective velocity
+w = Geometry.WVector.(speed .* ones(FT, fs))
+
+# Define a pulse wave or square wave
+pulse(z, t, z₀, zₕ, z₁) = abs(z.z - speed * t) ≤ zₕ ? z₁ : z₀
```
Comment (valeriabarra) [inline examples/column/zalesak_fct_advection.jl:102]: I would generalize this (and the comment above), for the case in which you want to run this for one (or more) periods

File: src/Operators/finitedifference.jl (hunk 5 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl:1801]: I think both A and 0 are correct, A->third order upwind boundary condition implementation, 0->first order upwind boundary condition implementation. I tried both, 0-> gives better results in terms of overshoting

File: examples/column/zalesak_fct_advection.jl (hunk 9 above)
Comment (Zhengyu-Huang) [reply examples/column/zalesak_fct_advection.jl:102]: I think the current setup is fine, I copy the test as your bb_fct test. I only run 1 period to understand the algorithm

File: docs/refs.bib (hunk 7 above)
Comment (Zhengyu-Huang) [reply docs/refs.bib:33]: just updated the citation

**Comment by Zhengyu-Huang** [comment]
@charleskawczynski could you help me with the unit tests? I saw the unit test failed at JET Test, do you know what that is?

File: src/Operators/finitedifference.jl [hunk 10]
```diff
@@ -1588,6 +1588,209 @@ end
...
+        Aⱼ₊₁₂ * (ϕⱼ₊₂ᵗᵈ - ϕⱼ₊₁ᵗᵈ) < stable_zero ||
+        Aⱼ₊₁₂ * (ϕⱼᵗᵈ - ϕⱼ₋₁ᵗᵈ) < stable_zero
+    )
+        Aⱼ₊₁₂ = stable_zero
+    end
+    ϕⱼᵐᵃˣ = max(ϕⱼ₋₁, ϕⱼ, ϕⱼ₊₁, ϕⱼ₋₁ᵗᵈ, ϕⱼᵗᵈ, ϕⱼ₊₁ᵗᵈ)
+    ϕⱼᵐⁱⁿ = min(ϕⱼ₋₁, ϕⱼ, ϕⱼ₊₁, ϕⱼ₋₁ᵗᵈ, ϕⱼᵗᵈ, ϕⱼ₊₁ᵗᵈ)
+    Pⱼ⁺ = max(stable_zero, Aⱼ₋₁₂) - min(stable_zero, Aⱼ₊₁₂)
+    Qⱼ⁺ = (ϕⱼᵐᵃˣ - ϕⱼᵗᵈ) * dxⱼ
+    Rⱼ⁺ = (Pⱼ⁺ > stable_zero ? min(1, Qⱼ⁺ / Pⱼ⁺) : stable_zero)
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:1669]: Can we also use `one(::Type)` instead of `1` here?

**Comment by charleskawczynski** [comment]
> @charleskawczynski could you help me with the unit tests? I saw the unit test failed at JET Test, do you know what that is?

If what you're talking about is the 1.8 unit tests, those are broken on main, so it's not related to this PR.

File: src/Operators/finitedifference.jl (hunk 10 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl]: just updated

**Comment by Zhengyu-Huang** [comment]
> > @charleskawczynski could you help me with the unit tests? I saw the unit test failed at JET Test, do you know what that is?
> 
> If what you're talking about is the 1.8 unit tests, those are broken on main, so it's not related to this PR.

should I merge this PR without touching it?

File: src/Operators/finitedifference.jl (hunk 5 above)
Comment (valeriabarra) [reply src/Operators/finitedifference.jl]: You're saying that we should use 0 also for the BB FCT. I'd like these BCs to be consistent for the two operators, unless there is a specific reason why they shouldn't be. Otherwise it gets confusing for people who might be looking into this in the future and we can't remember why we made one choice over the other.

**Review by charleskawczynski — APPROVED** [review]

**Comment by valeriabarra** [comment]
@Zhengyu-Huang if you can, please see my question above. I think it's better to be consistent.

File: src/Operators/finitedifference.jl (hunk 5 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl]: I am not sure whether we should use 0 also for the BB FCT, could you try both for more tests to see which one is better? For Zalesak, using 0 reduces undershooting

**Comment by Zhengyu-Huang** [comment]
bors try

File: src/Operators/finitedifference.jl [hunk 11]
```diff
@@ -1588,6 +1588,210 @@ end
...
+- a center-valued field `Φ`
+- a center-valued field `Φᵗᵈ`
+```math
+Φ_j^{n+1} = Φ_j^{td} - \frac{dt}{dx}(C_{j+\frac{1}{2}}A_{j+\frac{1}{2}} - C_{j-\frac{1}{2}}A_{j-\frac{1}{2}})
+```
+This stencil is based on [zalesak1979fully](@cite).
+
+Supported boundary conditions are:
+- [`ThirdOrderOneSided(x₀)`](@ref): uses the third-order downwind reconstruction to compute `x` on the left boundary,
+and the third-order upwind reconstruction to compute `x` on the right boundary.
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:1610]: If in this case, the boundary condition is falling back to the first-order-upwinding case, then this docstring should be updated

Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl]: thanks! just change it to FirstOrder

File: src/Operators/finitedifference.jl [hunk 12]
```diff
@@ -1588,6 +1588,210 @@ end
...
+    idx,
+    hidx,
+    A_space,
+    Φ_space,
+    Φᵗᵈ_space,
+)
+    face_space = axes(A_space)
+    @assert idx <= left_face_boundary_idx(face_space) + 1
+
+    return Geometry.Contravariant3Vector(zero(eltype(eltype(A_space))))
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:1774]: Why two `eltype`s?

File: src/Operators/finitedifference.jl [hunk 13]
```diff
@@ -1588,6 +1588,210 @@ end
...
+    idx,
+    hidx,
+    A_space,
+    Φ_space,
+    Φᵗᵈ_space,
+)
+    face_space = axes(A_space)
+    @assert idx <= right_face_boundary_idx(face_space) - 1
+
+    return Geometry.Contravariant3Vector(zero(eltype(eltype(A_space))))
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:1790]: Why two `eltype`s?

File: src/Operators/finitedifference.jl (hunk 12 above)
Comment (Zhengyu-Huang) [reply src/Operators/finitedifference.jl:1774]: @simonbyrne  helps me with this, I guess the first one returns a vector, and the second returns a scalar?
